src/m3/ecore/importer.ts

```diff
@@ -0,0 +1,119 @@
+import {
+    Concept,
+    Feature,
+    FeaturesContainer,
+    Metamodel,
+    MetamodelElement,
+    PrimitiveType
+} from "../types.ts"
+import {MetamodelFactory} from "../factory.ts"
+import {hashingIdGen} from "../../id-generation.ts"
+import {
+    asArray,
+    EClassifier,
+    EcoreXml,
+    EStructuralFeature
+} from "./types.ts"
+import {ConceptType} from "../functions.ts"
+
+
+const refPrefix = "#//"
+const deref = (typeDescriptor: string): string =>
+    typeDescriptor.startsWith(refPrefix)
+        ? typeDescriptor.substring(refPrefix.length)
+        : typeDescriptor
+
+export const asLIonCoreMetamodel = (ecoreXml: EcoreXml): Metamodel => {
+    const ePackage = ecoreXml["ecore:EPackage"]
+    // TODO  an Ecore XML can contain multiple EPackage-s
+    const factory = new MetamodelFactory(ePackage["@name"], hashingIdGen({ checkForUniqueHash: true }))
+
+    // TODO  obtain from a built-in, imported M3 instance
+    const stringDatatype = factory.primitiveType("String")
+    const booleanDatatype = factory.primitiveType("boolean")
+    const intDatatype = factory.primitiveType("int")
+
+    const convertEDataType = (eDataType: string): PrimitiveType => {
+        switch (eDataType) {
+            case "ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString":
+            case "ecore:EDataType http://www.eclipse.org/emf/2003/XMLType#//String":
+                return stringDatatype
+            case "ecore:EDataType http://www.eclipse.org/emf/2003/XMLType#//Boolean":
+                return booleanDatatype
+            case "ecore:EDataType http://www.eclipse.org/emf/2003/XMLType#//Int":
+                return intDatatype
+            default:
+                throw new Error(`don't know what to convert this EDataType ref. descriptor to: ${eDataType}`)
+        }
+    }
+
+    const convertEClassifier = (eClassifier: EClassifier): ConceptType =>
+        factory.concept(eClassifier["@name"], false)
+    // TODO  ConceptInterface, Enumeration
+
+    const convertEStructuralFeature = (container: FeaturesContainer, feature: EStructuralFeature): Feature => {
+        const metaType = feature["@xsi:type"]
+        const name = feature["@name"]
+        switch (metaType) {
+            case "ecore:EAttribute": {
+                const property = factory.property(container, name)
+                    .ofType(convertEDataType(feature["@eType"]))
+                if (feature["@lowerBound"] === "0") {
+                    property.isOptional()
+                }
+                return property
+            }
+            case "ecore:EReference": {
+                const link = (
+                        feature["@containment"]
+                            ? factory.containment(container, name)
```

Comment:
What I do not understand is what happens in case of a reference to a Classifier we have not yet unserialized: do we create an unresolved reference which is solved later?

Reply:
It works because I first convert all EClassifiers but don't convert their EStructuralFeatures yet. I'll swap the functions around to try and make this a bit clearer.